### PR TITLE
add ntp setup service

### LINKED ISSUE #2219: More robust NTP server name resolution
While investigating a [case](https://github.com/threefoldtech/zos/issues/2194) of farmer's nodes not working properly due to failure to reach NTP servers, I landed on an interesting [thread](https://bbs.archlinux.org/viewtopic.php?id=292688) discussing the root cause of the issue.

The summary is this:

* Zos resolves `[0-3].pool.ntp.org` to find the IP of an NTP server to use
* ntp.org is using geolocation on the source IP of these requests and mapping them to regional server pools
* The farmer with the issue is located in the Philippines, and the pool of servers provided to the farmer's nodes is polluted with faulty servers (this is a country specific pool that can be queried directly at `ph.pool.ntp.org`)
* Until the ntp.org pool is cleaned up, this farmer and potentially any others in the Philippines will be liable to have time sync issues with Zos

I think we can make some simple changes to improve the robustness of Zos in reaching a functional NTP server. Maybe a three tiered approach:

1. What we do now, query `[0-3].pool.ntp.org`
2. Assuming we have already completed geoip, query the appropriate [continent pool](https://www.ntppool.org/zone/@) based on the node's location. For example `[0-3].asia.pool.ntp.org`
3. Go round robin on the remaining continent pools, if the local one has failed

That said, I am just now reading this on the ntp.org [vendors page](https://www.ntppool.org/vendors.html):

>To allow you to use the pool as the default time service in your application, we will set you up with special hostnames, for example 0.vendor.pool.ntp.org, 1.vendor.pool.ntp.org, 2.vendor.pool.ntp.org and 3.vendor.pool.ntp.org.
>
> You **must absolutely not use the default pool.ntp.org zone names** as the default configuration in your application or appliance.

Pretty sure that's meant to include us, and it would explain why all the distros have their own subdomains (`arch.pool.ntp.org`, `ubuntu.pool.ntp.org`, etc). But I think it leaves us with the same issue, as these subdomains seem to resolve in the same way as the base domain.

I'm happy to pursue getting our own subdomain from ntp.org. I can also ask while I'm at it what their recommended approach regarding this issue would be.

### PR DESCRIPTION
**Pull request by Omarabdul3ziz**
- fix https://github.com/threefoldtech/zos/issues/2219

### CODE REVIEW

File: config/etc/zinit/ntp.yaml [hunk 1]
```diff
@@ -1 +1 @@
-exec: ntpd -n
+exec: sh /etc/zinit/init/ntpd-setup.sh
```
Comment (muhamadazmy) [inline config/etc/zinit/ntp.yaml]: I think now `ntpd-setup` is not the best name. Just call it `ntpd.sh`

File: config/etc/zinit/init/ntpd-setup.sh [hunk 2]
```diff
@@ -0,0 +1,19 @@
...
+
+ntp_flags=$(cat /proc/cmdline | grep -o 'ntp=.*')
+
+if [ -n "$ntp_flags" ]; then
+  ntp_flags=${ntp_flags#*=}
+
+  temp_file=$(mktemp)
+
+  for server in $(echo "$ntp_flags" | tr ',' ' '); do
+    echo "server $server" >> $temp_file
```
Comment (muhamadazmy) [inline config/etc/zinit/init/ntpd-setup.sh]: Why do you need a temp file, can't you just write the content directly to the ntp.conf file?

Comment (muhamadazmy) [reply config/etc/zinit/init/ntpd-setup.sh]: you first clear the file (either remove, or truncate) then in this loop just do the echo directly to the config file.

Comment (maxux) [reply config/etc/zinit/init/ntpd-setup.sh:11]: Or you can just append `-p PEER` (can be repeated) :p

Comment (maxux) [reply config/etc/zinit/init/ntpd-setup.sh]: Or you can just append `-p PEER` (can be repeated) :p

```sh
exec ntpd -n $(echo "-p $ntp_flags" | sed s/,/' -p '/g)
```
Should just works if `$ntp_flag`  is set.

Comment (Omarabdul3ziz) [reply config/etc/zinit/init/ntpd-setup.sh]: - > Or you can just append -p PEER (can be repeated) :p

  I checked the `ntpd` docs and looks `-p` flag is for providing `pidfile` not a peer server
  image


- > you first clear the file (either remove, or truncate) then in this loop just do the echo directly to the config file.
  
  i didn't want to clear the file until i am sure there is content that will override the existing one. providing an empty flag like `ntp=` will empty the file with no new content. but yeah not the best approach.

  i modified the script with a better check for the content of `ntp_flags` and got rid of the tmp file.

Comment (maxux) [reply config/etc/zinit/init/ntpd-setup.sh]: Which ntpd doc did you read ? We use the ntpd from busybox.

Comment (muhamadazmy) [reply config/etc/zinit/init/ntpd-setup.sh]: @maxux 
I think using the `-p` flag is a good idea. And then we don't need to even update the file.

Comment (Omarabdul3ziz) [reply config/etc/zinit/init/ntpd-setup.sh]: > Which ntpd doc did you read ? We use the ntpd from busybox.

i think i was reading another ntpd docs. checked the one on zos and it do use `-p` for peers

updated the code to use -p as suggested

File: config/etc/zinit/init/ntpd.sh [hunk 3]
```diff
@@ -0,0 +1,10 @@
+#!/bin/sh
+
+ntp_flags=$(grep -o 'ntp=.*' /proc/cmdline | sed 's/^ntp=//')
+
+params=""
+if [ -n "$ntp_flags" ]; then
+  params=$(echo "-p $ntp_flags" | sed s/,/' -p '/g)
```
Comment (muhamadazmy) [inline config/etc/zinit/init/ntpd.sh:7]: I am wondering if people can do something like 
```
ntp="\" && echo 'key' > /root/.ssh/authorized_keys && exit 0"
```
to inject his key into ssh authorized keys via command line.

**Review by muhamadazmy — APPROVED** [review]

**Review by muhamadazmy — APPROVED** [review]

**Pull request merged**